{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "mount_file_id": "1EE8ToMNZVkhIRr6N6eiUAO9sxYAe6QwX",
      "authorship_tag": "ABX9TyNzvKhQi0i3ZVm5hYHXpLTf",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PRISM-SING-HEALTH/2025-NCP-Projects/blob/main/NCP_Variant_DB.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 381,
      "metadata": {
        "id": "0_QRs2CDqtdq"
      },
      "outputs": [],
      "source": [
        "# Initialise imports\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "\n",
        "# Base path for excel files\n",
        "base_path = '/content/drive/MyDrive/Variant_DB/Mock_Shared_Drive/'\n",
        "local_path = '/content/drive/MyDrive/Variant_DB/Mock_Local_Drive/'\n",
        "\n",
        "# Define standard columns\n",
        "standard_columns = [\n",
        "    'MRN',\n",
        "    'Patient Name',\n",
        "    'Phenotype',\n",
        "    'Solved Status',\n",
        "    'Gene',\n",
        "    'Transcript',\n",
        "    'Variant',\n",
        "    'HGVSg',\n",
        "    'HGVSc',\n",
        "    'HGVSp'\n",
        "]"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Importing data into dataframes\n",
        "print('Importing data...')\n",
        "\n",
        "try:\n",
        "  # Lab Cases\n",
        "  lab_cases_df = pd.read_excel(\n",
        "      base_path + '/Internal_Drive/Lab_Case_list.xlsx',\n",
        "      sheet_name = 'Sheet1',\n",
        "      header = 0,\n",
        "      usecols = 'F, AF:AI, AL'\n",
        "  )\n",
        "  print('Passed.')\n",
        "\n",
        "  # ATM Summary (fix)\n",
        "  atm_summary_df = pd.read_excel(\n",
        "      base_path + '/Internal_Drive/ATM_c.3576G~A_p.Lys1192Lys.xlsx',\n",
        "      sheet_name='SUMMARY',\n",
        "      header=None,  # Assuming no headers in the Excel sheet\n",
        "      usecols='A:B'  # Only read columns A and B\n",
        "  )\n",
        "  # Additional filtering due to unconventional formatting\n",
        "  print('Performing additional filtering...')\n",
        "  atm_summary_df_filtered = atm_summary_df.iloc[9:16]  # Ensure rows are correctly indexed\n",
        "  atm_summary_df_filtered.reset_index(drop=True, inplace=True)  # Reset index for clean output\n",
        "\n",
        "  # Convert to a structured DataFrame\n",
        "  print('Transforming into structured DataFrame...')\n",
        "  atm_structured_df = atm_summary_df_filtered.set_index(0).T  # Set column 0 as the header, transpose\n",
        "  atm_structured_df.reset_index(drop=True, inplace=True)  # Reset index for clean final DataFrame\n",
        "\n",
        "  # Invitae Summary\n",
        "  invitae_summary_df = pd.read_excel(\n",
        "      base_path + '/External_Drive/Invitae_List.xlsx',\n",
        "      sheet_name = 'Invitae list header',\n",
        "      header = 0,\n",
        "      usecols = 'E:F, M:T'\n",
        "  )\n",
        "  print('Passed.')\n",
        "\n",
        "  # Clinical Summary\n",
        "  clinical_summary_df = pd.read_excel(\n",
        "      base_path + '/Internal_Drive/Clinic_List.xlsx',\n",
        "      sheet_name = '11 Dec',\n",
        "      header = 0,\n",
        "      usecols = 'D:E, K')\n",
        "  print('Passed.')\n",
        "\n",
        "  # Research Summary\n",
        "  research_summary_df = pd.read_excel(\n",
        "      base_path + '/Research_Drive/Research_List.xlsx',\n",
        "      sheet_name = 'Overall List',\n",
        "      header = 2,\n",
        "      usecols = 'B, W:X, AF, AS:AT, AV:AW')\n",
        "  print('Passed.')\n",
        "\n",
        "except FileNotFoundError as e:\n",
        "  print(f\"File not found: {e}\")\n",
        "except ValueError as e:\n",
        "  print(f\"Error reading sheet or invalid data format: {e}\")\n",
        "\n",
        "print('Data import complete.')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QCnw8yNvEMj_",
        "outputId": "f46a2714-62d9-4c2a-b8f0-bde51bef4de6"
      },
      "execution_count": 382,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Importing data...\n",
            "Passed.\n",
            "Performing additional filtering...\n",
            "Transforming into structured DataFrame...\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n",
            "Data import complete.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Print dataframes\n",
        "print('Printing dataframes...')\n",
        "\n",
        "print('\\nLab Cases')\n",
        "print(lab_cases_df)\n",
        "\n",
        "print('\\nATM Summary')\n",
        "print(atm_structured_df)\n",
        "\n",
        "print('\\nInvitae Summary')\n",
        "print(invitae_summary_df)\n",
        "\n",
        "print('\\nClinical Summary')\n",
        "print(clinical_summary_df)\n",
        "\n",
        "print('\\nResearch Summary')\n",
        "print(research_summary_df)"
      ],
      "metadata": {
        "id": "1TE9tZBvhLKI",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "928a0b17-2b54-4dff-898a-0fc11d256e19"
      },
      "execution_count": 383,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Printing dataframes...\n",
            "\n",
            "Lab Cases\n",
            "                                           Phenotype          Variant_1_gene  \\\n",
            "0  Developmental / Behavioural - Hyperactivity; N...  chr6q14.2q15 (3.72 Mb)   \n",
            "1  Developmental / Behavioural - Global developme...                   PRRT2   \n",
            "2  Developmental / Behavioural - Global developme...                  STING1   \n",
            "3  Developmental / Behavioural - Intellectual dis...                     NaN   \n",
            "4  Developmental / Behavioural - Delayed fine mot...                   NTRK1   \n",
            "5  Developmental / Behavioural - Global developme...                  DYRK1A   \n",
            "6  Developmental / Behavioural - Delayed fine mot...                     NaN   \n",
            "\n",
            "                        Variant_1_HGVSg          Variant_1_HGVSc  \\\n",
            "0  NC_000006.12:g. 83707798_87418416del                      NaN   \n",
            "1                   chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2           NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3                                   NaN                      NaN   \n",
            "4                   chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "5                   chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "6                                   NaN                      NaN   \n",
            "\n",
            "                 Variant_1_HGVSp        Variant_1_Validation_Status  \n",
            "0                            NaN  Confirmed present by clinical CMA  \n",
            "1  NP_660282.2:p.Arg217ProfsTer8                  Confirmed present  \n",
            "2        NP_938023.1:p.Val155Met                  Confirmed present  \n",
            "3                            NaN                                NaN  \n",
            "4        NP_002520.2:p.Cys656Gly                  Confirmed present  \n",
            "5     NP_001334650.1:p.Trp146Ter                  Confirmed present  \n",
            "6                            NaN                                NaN  \n",
            "\n",
            "ATM Summary\n",
            "0     HGVS_Genomic_GRCh38/hg38 HGVS_MANE Select_Transcript_RefSeq  \\\n",
            "0  NC_000011.10:g.108281168G>A                        NM_000051.3   \n",
            "\n",
            "0 HGVS_MANE Select_cDNA HGVS_MANE Select_protein HGVS_MANE Select_Exon  \\\n",
            "0             c.3576G>A             p.(Lys1192=)                 24/53   \n",
            "\n",
            "0 HGVS_MANE Select_Molecular consequence    Zygosity  \n",
            "0                             Synonymous  Homozygous  \n",
            "\n",
            "Invitae Summary\n",
            "   Patient ID (MRN) Patient Name    Result  Gene   Transcript    Variant  \\\n",
            "0           43234.0    John Pork  POSITIVE  TSC2  NM_000548.3  c.1513C>T   \n",
            "1               NaN          NaN  NEGATIVE     -            -          -   \n",
            "\n",
            "  Protein Change Intron/Exon      Zygosity Classification  \n",
            "0      p.Arg505*     Exon 15  heterozygous     PATHOGENIC  \n",
            "1              -         NaN             -              -  \n",
            "\n",
            "Clinical Summary\n",
            "  Patient Name  Identification No.  \\\n",
            "0      Jon Doe               420.0   \n",
            "1          NaN                 NaN   \n",
            "\n",
            "                            Medical Prob description  \n",
            "0  Shwachman-Diamond Syndrome\\nloose stools- fat ...  \n",
            "1  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...  \n",
            "\n",
            "Research Summary\n",
            "  AUTO STATUS  IC No (MRN)        Name  \\\n",
            "0      Solved       1234.0  John Smith   \n",
            "1      Solved          NaN         NaN   \n",
            "\n",
            "                                           Phenotype Candidate gene (1)  \\\n",
            "0                   GDD with regression, Autism, FTT              CDK13   \n",
            "1  Dysmorphism, facial dysmorphism: arched eyebro...              KMT2D   \n",
            "\n",
            "  Transcript (1)     cDNA (1)  Protein (1)  \n",
            "0    NM_003718.5    c.2389G>C  p.Asp797His  \n",
            "1    NM_003482.3  c.12063dupG    p.T4022fs  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Standardise columns across dataframes\n",
        "print('Standardising columns across dataframes...')\n",
        "\n",
        "# Lab Cases\n",
        "lab_cases_cols = {\n",
        "    'Number variants detected'    :'Var Count',\n",
        "    'Variant_1_gene'              :'Gene',\n",
        "    'Variant_1_HGVSg'             :'HGVSg',\n",
        "    'Variant_1_HGVSc'             :'HGVSc',\n",
        "    'Variant_1_HGVSp'             :'HGVSp',\n",
        "    'Variant_1_zygosity'          :'Zygosity',\n",
        "    'Variant_1_inheritance'       :'Inheritance',\n",
        "    'Variant_1_Validation_Status' :'Solved Status'\n",
        "}\n",
        "\n",
        "lab_cases_df.rename(columns = lab_cases_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# ATM Summary\n",
        "atm_structured_df_cols = {\n",
        "    'HGVS_Genomic_GRCh38/hg38'            :'HSVSg',\n",
        "    'HGVS_MANE Select_Transcript_RefSeq'  :'Transcript',\n",
        "    'HGVS_MANE Select_cDNA'               :'HGVSc',\n",
        "    'HGVS_MANE Select_protein'            :'HGVSp',\n",
        "    'HUGO gene symbol'                    :'Gene',\n",
        "}\n",
        "\n",
        "atm_structured_df.rename(columns = atm_structured_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# Invitae Summary\n",
        "invitae_summary_df_cols = {\n",
        "    'Patient ID (MRN)'  :'MRN',\n",
        "    'Patient Name'      :'Patient Name',\n",
        "    'Result'            :'Solved Status',\n",
        "    'Gene'              :'Gene',\n",
        "    'Transcript'        :'Transcript',\n",
        "    'Variant'           :'Variant',\n",
        "    'HGVSc'             :'HGVSc',\n",
        "    'Protein Change'    :'HGVSp',\n",
        "}\n",
        "\n",
        "invitae_summary_df.rename(columns = invitae_summary_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# Clinical Summary\n",
        "clinical_summary_df_cols = {\n",
        "    'Identification No.'        :'MRN',\n",
        "    'Medical Prob description'  :'Phenotype',\n",
        "    'Patient Name'              :'Patient Name'\n",
        "}\n",
        "\n",
        "clinical_summary_df.rename(columns = clinical_summary_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# Research Summary\n",
        "research_summary_df_cols = {\n",
        "    'IC No (MRN)'         :'MRN',\n",
        "    'Name'                :'Patient Name',\n",
        "    'Candidate gene (1)'  :'Gene',\n",
        "    'Transcript (1)'      :'Transcript',\n",
        "    'cDNA (1)'            :'HGVSc',\n",
        "    'Protein (1)'         :'HGVSp',\n",
        "    'AUTO STATUS'         :'Solved Status'\n",
        "}\n",
        "\n",
        "research_summary_df.rename(columns = research_summary_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DLRt0qwiIKBZ",
        "outputId": "c6e2dd87-ffca-46e7-bb37-5a7f6be7cd2f"
      },
      "execution_count": 384,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Standardising columns across dataframes...\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Print dataframes after renaming\n",
        "print('Printing dataframes...')\n",
        "\n",
        "print('\\nLab Cases')\n",
        "print(lab_cases_df)\n",
        "\n",
        "print('\\nATM Summary')\n",
        "print(atm_structured_df)\n",
        "\n",
        "print('\\nInvitae Summary')\n",
        "print(invitae_summary_df)\n",
        "\n",
        "print('\\nClinical Summary')\n",
        "print(clinical_summary_df)\n",
        "\n",
        "print('\\nResearch Summary')\n",
        "print(research_summary_df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6x-Sz1WP8kvV",
        "outputId": "479a8783-b02b-4720-a869-1b92cd44e6ac"
      },
      "execution_count": 385,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Printing dataframes...\n",
            "\n",
            "Lab Cases\n",
            "                                           Phenotype                    Gene  \\\n",
            "0  Developmental / Behavioural - Hyperactivity; N...  chr6q14.2q15 (3.72 Mb)   \n",
            "1  Developmental / Behavioural - Global developme...                   PRRT2   \n",
            "2  Developmental / Behavioural - Global developme...                  STING1   \n",
            "3  Developmental / Behavioural - Intellectual dis...                     NaN   \n",
            "4  Developmental / Behavioural - Delayed fine mot...                   NTRK1   \n",
            "5  Developmental / Behavioural - Global developme...                  DYRK1A   \n",
            "6  Developmental / Behavioural - Delayed fine mot...                     NaN   \n",
            "\n",
            "                                  HGVSg                    HGVSc  \\\n",
            "0  NC_000006.12:g. 83707798_87418416del                      NaN   \n",
            "1                   chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2           NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3                                   NaN                      NaN   \n",
            "4                   chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "5                   chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "6                                   NaN                      NaN   \n",
            "\n",
            "                           HGVSp                      Solved Status  \n",
            "0                            NaN  Confirmed present by clinical CMA  \n",
            "1  NP_660282.2:p.Arg217ProfsTer8                  Confirmed present  \n",
            "2        NP_938023.1:p.Val155Met                  Confirmed present  \n",
            "3                            NaN                                NaN  \n",
            "4        NP_002520.2:p.Cys656Gly                  Confirmed present  \n",
            "5     NP_001334650.1:p.Trp146Ter                  Confirmed present  \n",
            "6                            NaN                                NaN  \n",
            "\n",
            "ATM Summary\n",
            "0                        HSVSg   Transcript      HGVSc         HGVSp  \\\n",
            "0  NC_000011.10:g.108281168G>A  NM_000051.3  c.3576G>A  p.(Lys1192=)   \n",
            "\n",
            "0 HGVS_MANE Select_Exon HGVS_MANE Select_Molecular consequence    Zygosity  \n",
            "0                 24/53                             Synonymous  Homozygous  \n",
            "\n",
            "Invitae Summary\n",
            "       MRN Patient Name Solved Status  Gene   Transcript    Variant  \\\n",
            "0  43234.0    John Pork      POSITIVE  TSC2  NM_000548.3  c.1513C>T   \n",
            "1      NaN          NaN      NEGATIVE     -            -          -   \n",
            "\n",
            "       HGVSp Intron/Exon      Zygosity Classification  \n",
            "0  p.Arg505*     Exon 15  heterozygous     PATHOGENIC  \n",
            "1          -         NaN             -              -  \n",
            "\n",
            "Clinical Summary\n",
            "  Patient Name    MRN                                          Phenotype\n",
            "0      Jon Doe  420.0  Shwachman-Diamond Syndrome\\nloose stools- fat ...\n",
            "1          NaN    NaN  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...\n",
            "\n",
            "Research Summary\n",
            "  Solved Status     MRN Patient Name  \\\n",
            "0        Solved  1234.0   John Smith   \n",
            "1        Solved     NaN          NaN   \n",
            "\n",
            "                                           Phenotype   Gene   Transcript  \\\n",
            "0                   GDD with regression, Autism, FTT  CDK13  NM_003718.5   \n",
            "1  Dysmorphism, facial dysmorphism: arched eyebro...  KMT2D  NM_003482.3   \n",
            "\n",
            "         HGVSc        HGVSp  \n",
            "0    c.2389G>C  p.Asp797His  \n",
            "1  c.12063dupG    p.T4022fs  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Query item testing\n",
        "\n",
        "#print('Initialising querying...\\n')\n",
        "\n",
        "#print('Test: Lab Cases\\n')\n",
        "#search_item = 'PRRT2'  # Item to query\n",
        "#from_category = 'Var1 Gene' # Category to query from\n",
        "#lab_cases_result = lab_cases_df[lab_cases_df[from_category] == search_item]\n",
        "#print(lab_cases_result)\n",
        "\n",
        "#search_item = 'PRRT2'  # Item to query\n",
        "#from_category = 'Var1 Gene' # Category to query from\n",
        "#atm_summary_result = lab_cases_df[lab_cases_df[from_category] == search_item]\n",
        "\n",
        "#print('Test: Invitae Cases\\n')\n",
        "#search_item = 'TSC2'  # Item to query\n",
        "#from_category = 'Gene' # Category to query from\n",
        "#invitae_summary_result = invitae_summary_df[invitae_summary_df[from_category] == search_item]\n",
        "#print(invitae_summary_result)\n",
        "\n",
        "#print('Test: Clinical Cases')\n",
        "#search_item = 'VSD'  # Item to query\n",
        "#from_category = 'Medical Prob Desc.' # Category to query from\n",
        "#clinical_summary_result = clinical_summary_df[clinical_summary_df[from_category] == search_item]\n",
        "#print(clinical_summary_result)  # ERROR: Empty dataframe\n",
        "\n",
        "#print('Test: Research Cases')\n",
        "#search_item = 'NM_003718.5'  # Item to query\n",
        "#from_category = 'Transcript (1)' # Category to query from\n",
        "#research_summary_result = research_summary_df[research_summary_df[from_category] == search_item]  # KeyERROR\n",
        "#print(research_summary_result)\n",
        "\n"
      ],
      "metadata": {
        "id": "u8dZt3bQKo7i"
      },
      "execution_count": 386,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Export results to excel\n",
        "#print('Exporting results...')\n",
        "#lab_cases_result.to_excel(local_path + 'lab_cases_results.xlsx', index = False)\n",
        "#clinical_summary_result.to_excel(local_path + 'clinical_summary_results.xlsx', index = False)\n",
        "#invitae_summary_result.to_excel(local_path + 'invitae_summary_results.xlsx', index = False)\n",
        "#clinical_summary_result.to_excel(local_path + 'clinical_summary_results.xlsx', index = False)\n",
        "#research_summary_result.to_excel(local_path + 'research_summary_results.xlsx', index = False)\n",
        "\n",
        "#print(\"Results exported successfully!\")"
      ],
      "metadata": {
        "id": "dz5s_DIsKz-Y"
      },
      "execution_count": 387,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Combining all 5 dataframes into a singluar dataframe.\n",
        "\n",
        "def standardise_dataframe(df, columns):\n",
        "  standardised_df = df.reindex(columns = columns, fill_value = np.nan)\n",
        "  return standardised_df\n",
        "\n",
        "print('Standardising all dataframes...')\n",
        "\n",
        "lab_cases_df_standard = standardise_dataframe(lab_cases_df, standard_columns)\n",
        "atm_structured_df_standard = standardise_dataframe(atm_structured_df, standard_columns)\n",
        "invitae_summary_df_standard = standardise_dataframe(invitae_summary_df, standard_columns)\n",
        "clinical_summary_df_standard = standardise_dataframe(clinical_summary_df, standard_columns)\n",
        "research_summary_df_standard = standardise_dataframe(research_summary_df, standard_columns)\n",
        "\n",
        "print('Standardisation complete.')\n",
        "\n",
        "print('Individual dataframes:')\n",
        "print('\\nLab Cases:')\n",
        "print(lab_cases_df_standard.head())\n",
        "\n",
        "print('\\nATM:')\n",
        "print(atm_structured_df_standard.head())\n",
        "\n",
        "print('\\nInvitae:')\n",
        "print(invitae_summary_df_standard.head())\n",
        "\n",
        "print('\\nClinical:')\n",
        "print(clinical_summary_df_standard.head())\n",
        "\n",
        "print('\\nResearch:')\n",
        "print(research_summary_df_standard.head())\n",
        "\n",
        "print('\\nCombining dataframes...')\n",
        "combined_df = pd.concat([\n",
        "    lab_cases_df_standard,\n",
        "    atm_structured_df_standard,\n",
        "    invitae_summary_df_standard,\n",
        "    clinical_summary_df_standard,\n",
        "    research_summary_df_standard\n",
        "], ignore_index = True)\n",
        "\n",
        "print('All dataframes combined successfully.')\n",
        "\n",
        "print('Combined dataframe:')\n",
        "print(combined_df)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xQd3sGenNIzt",
        "outputId": "a9a9d0b5-3221-4a50-a186-51b93dffd940"
      },
      "execution_count": 388,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Standardising all dataframes...\n",
            "Standardisation complete.\n",
            "Individual dataframes:\n",
            "\n",
            "Lab Cases:\n",
            "   MRN  Patient Name                                          Phenotype  \\\n",
            "0  NaN           NaN  Developmental / Behavioural - Hyperactivity; N...   \n",
            "1  NaN           NaN  Developmental / Behavioural - Global developme...   \n",
            "2  NaN           NaN  Developmental / Behavioural - Global developme...   \n",
            "3  NaN           NaN  Developmental / Behavioural - Intellectual dis...   \n",
            "4  NaN           NaN  Developmental / Behavioural - Delayed fine mot...   \n",
            "\n",
            "                       Solved Status                    Gene  Transcript  \\\n",
            "0  Confirmed present by clinical CMA  chr6q14.2q15 (3.72 Mb)         NaN   \n",
            "1                  Confirmed present                   PRRT2         NaN   \n",
            "2                  Confirmed present                  STING1         NaN   \n",
            "3                                NaN                     NaN         NaN   \n",
            "4                  Confirmed present                   NTRK1         NaN   \n",
            "\n",
            "   Variant                                 HGVSg                  HGVSc  \\\n",
            "0      NaN  NC_000006.12:g. 83707798_87418416del                    NaN   \n",
            "1      NaN                   chr16:g.29813703dup   NM_145239.3:c.649dup   \n",
            "2      NaN           NC_000005.10:g.139480847C>T   NM_198282.4:c.463G>A   \n",
            "3      NaN                                   NaN                    NaN   \n",
            "4      NaN                   chr1:g.156879282T>G  NM_002529.4:c.1966T>G   \n",
            "\n",
            "                           HGVSp  \n",
            "0                            NaN  \n",
            "1  NP_660282.2:p.Arg217ProfsTer8  \n",
            "2        NP_938023.1:p.Val155Met  \n",
            "3                            NaN  \n",
            "4        NP_002520.2:p.Cys656Gly  \n",
            "\n",
            "ATM:\n",
            "0  MRN  Patient Name  Phenotype  Solved Status  Gene   Transcript  Variant  \\\n",
            "0  NaN           NaN        NaN            NaN   NaN  NM_000051.3      NaN   \n",
            "\n",
            "0  HGVSg      HGVSc         HGVSp  \n",
            "0    NaN  c.3576G>A  p.(Lys1192=)  \n",
            "\n",
            "Invitae:\n",
            "       MRN Patient Name  Phenotype Solved Status  Gene   Transcript  \\\n",
            "0  43234.0    John Pork        NaN      POSITIVE  TSC2  NM_000548.3   \n",
            "1      NaN          NaN        NaN      NEGATIVE     -            -   \n",
            "\n",
            "     Variant  HGVSg  HGVSc      HGVSp  \n",
            "0  c.1513C>T    NaN    NaN  p.Arg505*  \n",
            "1          -    NaN    NaN          -  \n",
            "\n",
            "Clinical:\n",
            "     MRN Patient Name                                          Phenotype  \\\n",
            "0  420.0      Jon Doe  Shwachman-Diamond Syndrome\\nloose stools- fat ...   \n",
            "1    NaN          NaN  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...   \n",
            "\n",
            "   Solved Status  Gene  Transcript  Variant  HGVSg  HGVSc  HGVSp  \n",
            "0            NaN   NaN         NaN      NaN    NaN    NaN    NaN  \n",
            "1            NaN   NaN         NaN      NaN    NaN    NaN    NaN  \n",
            "\n",
            "Research:\n",
            "      MRN Patient Name                                          Phenotype  \\\n",
            "0  1234.0   John Smith                   GDD with regression, Autism, FTT   \n",
            "1     NaN          NaN  Dysmorphism, facial dysmorphism: arched eyebro...   \n",
            "\n",
            "  Solved Status   Gene   Transcript  Variant  HGVSg        HGVSc        HGVSp  \n",
            "0        Solved  CDK13  NM_003718.5      NaN    NaN    c.2389G>C  p.Asp797His  \n",
            "1        Solved  KMT2D  NM_003482.3      NaN    NaN  c.12063dupG    p.T4022fs  \n",
            "\n",
            "Combining dataframes...\n",
            "All dataframes combined successfully.\n",
            "Combined dataframe:\n",
            "        MRN Patient Name                                          Phenotype  \\\n",
            "0       NaN          NaN  Developmental / Behavioural - Hyperactivity; N...   \n",
            "1       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "2       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "3       NaN          NaN  Developmental / Behavioural - Intellectual dis...   \n",
            "4       NaN          NaN  Developmental / Behavioural - Delayed fine mot...   \n",
            "5       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "6       NaN          NaN  Developmental / Behavioural - Delayed fine mot...   \n",
            "7       NaN          NaN                                                NaN   \n",
            "8   43234.0    John Pork                                                NaN   \n",
            "9       NaN          NaN                                                NaN   \n",
            "10    420.0      Jon Doe  Shwachman-Diamond Syndrome\\nloose stools- fat ...   \n",
            "11      NaN          NaN  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...   \n",
            "12   1234.0   John Smith                   GDD with regression, Autism, FTT   \n",
            "13      NaN          NaN  Dysmorphism, facial dysmorphism: arched eyebro...   \n",
            "\n",
            "                        Solved Status                    Gene   Transcript  \\\n",
            "0   Confirmed present by clinical CMA  chr6q14.2q15 (3.72 Mb)          NaN   \n",
            "1                   Confirmed present                   PRRT2          NaN   \n",
            "2                   Confirmed present                  STING1          NaN   \n",
            "3                                 NaN                     NaN          NaN   \n",
            "4                   Confirmed present                   NTRK1          NaN   \n",
            "5                   Confirmed present                  DYRK1A          NaN   \n",
            "6                                 NaN                     NaN          NaN   \n",
            "7                                 NaN                     NaN  NM_000051.3   \n",
            "8                            POSITIVE                    TSC2  NM_000548.3   \n",
            "9                            NEGATIVE                       -            -   \n",
            "10                                NaN                     NaN          NaN   \n",
            "11                                NaN                     NaN          NaN   \n",
            "12                             Solved                   CDK13  NM_003718.5   \n",
            "13                             Solved                   KMT2D  NM_003482.3   \n",
            "\n",
            "      Variant                                 HGVSg                    HGVSc  \\\n",
            "0         NaN  NC_000006.12:g. 83707798_87418416del                      NaN   \n",
            "1         NaN                   chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2         NaN           NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3         NaN                                   NaN                      NaN   \n",
            "4         NaN                   chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "5         NaN                   chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "6         NaN                                   NaN                      NaN   \n",
            "7         NaN                                   NaN                c.3576G>A   \n",
            "8   c.1513C>T                                   NaN                      NaN   \n",
            "9           -                                   NaN                      NaN   \n",
            "10        NaN                                   NaN                      NaN   \n",
            "11        NaN                                   NaN                      NaN   \n",
            "12        NaN                                   NaN                c.2389G>C   \n",
            "13        NaN                                   NaN              c.12063dupG   \n",
            "\n",
            "                            HGVSp  \n",
            "0                             NaN  \n",
            "1   NP_660282.2:p.Arg217ProfsTer8  \n",
            "2         NP_938023.1:p.Val155Met  \n",
            "3                             NaN  \n",
            "4         NP_002520.2:p.Cys656Gly  \n",
            "5      NP_001334650.1:p.Trp146Ter  \n",
            "6                             NaN  \n",
            "7                    p.(Lys1192=)  \n",
            "8                       p.Arg505*  \n",
            "9                               -  \n",
            "10                            NaN  \n",
            "11                            NaN  \n",
            "12                    p.Asp797His  \n",
            "13                      p.T4022fs  \n"
          ]
        }
      ]
    }
  ]
}