{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PRISM-SING-HEALTH/2025-NCP-Projects/blob/main/NCP_Variant_DB.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import yaml"
      ],
      "metadata": {
        "id": "Z-woNJnevcob"
      },
      "execution_count": 241,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def load_config(config_file):\n",
        "  \"\"\"\n",
        "  Loads configuration from YAML file.\n",
        "\n",
        "  Args:\n",
        "    config_file (str): Path to the YAML configuration file.\n",
        "\n",
        "  Returns:\n",
        "    dict: Configuration settings.\n",
        "  \"\"\"\n",
        "\n",
        "  with open(config_file, 'r') as file:\n",
        "    config = yaml.safe_load(file)\n",
        "  return config\n",
        "\n",
        "# Load configuration\n",
        "config = load_config('/content/drive/MyDrive/Variant_DB/Mock_Local_Drive/config.yaml')\n",
        "\n",
        "# Initialise file paths for excel files\n",
        "base_path = config['base_path']\n",
        "local_path = config['local_path']\n",
        "files = config['files']"
      ],
      "metadata": {
        "id": "0JCnmuPNvniD"
      },
      "execution_count": 242,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": 243,
      "metadata": {
        "id": "0_QRs2CDqtdq"
      },
      "outputs": [],
      "source": [
        "# Define standard columns\n",
        "standard_columns = [\n",
        "    'MRN',\n",
        "    'Patient Name',\n",
        "    'Phenotype',\n",
        "    'Solved Status',\n",
        "    'Gene',\n",
        "    'Transcript',\n",
        "    'Variant',\n",
        "    'HGVSg',\n",
        "    'HGVSc',\n",
        "    'HGVSp'\n",
        "]"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 244,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "QCnw8yNvEMj_",
        "outputId": "30089f8d-2c98-4fc5-c5a6-8ad3820054b8"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Importing data...\n",
            "Passed.\n",
            "Performing additional filtering...\n",
            "Transforming into structured DataFrame...\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n",
            "Data import complete.\n"
          ]
        }
      ],
      "source": [
        "# Importing data into dataframes\n",
        "print('Importing data...')\n",
        "try:\n",
        "  # Lab Cases\n",
        "  lab_cases_df = pd.read_excel(\n",
        "      f\"{base_path}{files['lab_cases']}\",\n",
        "      sheet_name = 'Sheet1',\n",
        "      header = 0,\n",
        "      usecols = 'F, AF:AI, AL'\n",
        "  )\n",
        "  print('Passed.')\n",
        "\n",
        "  # ATM Summary (fix)\n",
        "  atm_summary_df = pd.read_excel(\n",
        "      f\"{base_path}{files['atm_summary']}\",\n",
        "      sheet_name='SUMMARY',\n",
        "      header = None,\n",
        "      usecols = 'A:B'\n",
        "  )\n",
        "\n",
        "  # Additional filtering due to unconventional formatting\n",
        "  print('Performing additional filtering...')\n",
        "  atm_summary_df_filtered = atm_summary_df.iloc[9:16]  # Ensure rows are correctly indexed\n",
        "  atm_summary_df_filtered.reset_index(drop=True, inplace=True)  # Reset index for clean output\n",
        "\n",
        "  # Convert to a structured DataFrame\n",
        "  print('Transforming into structured DataFrame...')\n",
        "  atm_structured_df = atm_summary_df_filtered.set_index(0).T  # Set column 0 as the header, transpose\n",
        "  atm_structured_df.reset_index(drop=True, inplace=True)  # Reset index for clean final DataFrame\n",
        "\n",
        "  # Invitae Summary\n",
        "  invitae_summary_df = pd.read_excel(\n",
        "      f\"{base_path}{files['invitae_summary']}\",\n",
        "      sheet_name = 'Invitae list header',\n",
        "      header = 0,\n",
        "      usecols = 'E:F, M:T'\n",
        "  )\n",
        "  print('Passed.')\n",
        "\n",
        "  # Clinical Summary\n",
        "  clinical_summary_df = pd.read_excel(\n",
        "      f\"{base_path}{files['clinical_summary']}\",\n",
        "      sheet_name = '11 Dec',\n",
        "      header = 0,\n",
        "      usecols = 'D:E, K')\n",
        "  print('Passed.')\n",
        "\n",
        "  # Research Summary\n",
        "  research_summary_df = pd.read_excel(\n",
        "      f\"{base_path}{files['research_summary']}\",\n",
        "      sheet_name = 'Overall List',\n",
        "      header = 2,\n",
        "      usecols = 'B, W:X, AF, AS:AT, AV:AW')\n",
        "  print('Passed.')\n",
        "\n",
        "except FileNotFoundError as e:\n",
        "  print(f\"File not found: {e}\")\n",
        "except ValueError as e:\n",
        "  print(f\"Error reading sheet or invalid data format: {e}\")\n",
        "\n",
        "print('Data import complete.')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 245,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "1TE9tZBvhLKI",
        "outputId": "bd2ed27a-22c9-4ca4-b329-73b3709af3d4"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Printing dataframes...\n",
            "\n",
            "Lab Cases\n",
            "                                           Phenotype Variant_1_gene  \\\n",
            "0  Developmental / Behavioural - Global developme...          PRRT2   \n",
            "1  Developmental / Behavioural - Global developme...          PRRT2   \n",
            "2  Developmental / Behavioural - Global developme...         STING1   \n",
            "3  Developmental / Behavioural - Global developme...         DYRK1A   \n",
            "4  Developmental / Behavioural - Delayed fine mot...          NTRK1   \n",
            "5  Developmental / Behavioural - Global developme...         DYRK1A   \n",
            "6  Developmental / Behavioural - Global developme...         DYRK1A   \n",
            "\n",
            "               Variant_1_HGVSg          Variant_1_HGVSc  \\\n",
            "0          chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "1          chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2  NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "4          chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "5          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "6          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "\n",
            "                 Variant_1_HGVSp Variant_1_Validation_Status  \n",
            "0  NP_660282.2:p.Arg217ProfsTer8           Confirmed present  \n",
            "1  NP_660282.2:p.Arg217ProfsTer8           Confirmed present  \n",
            "2        NP_938023.1:p.Val155Met           Confirmed present  \n",
            "3     NP_001334650.1:p.Trp146Ter           Confirmed present  \n",
            "4        NP_002520.2:p.Cys656Gly           Confirmed present  \n",
            "5     NP_001334650.1:p.Trp146Ter           Confirmed present  \n",
            "6     NP_001334650.1:p.Trp146Ter           Confirmed present  \n",
            "\n",
            "ATM Summary\n",
            "0     HGVS_Genomic_GRCh38/hg38 HGVS_MANE Select_Transcript_RefSeq  \\\n",
            "0  NC_000011.10:g.108281168G>A                        NM_000051.3   \n",
            "\n",
            "0 HGVS_MANE Select_cDNA HGVS_MANE Select_protein HGVS_MANE Select_Exon  \\\n",
            "0             c.3576G>A             p.(Lys1192=)                 24/53   \n",
            "\n",
            "0 HGVS_MANE Select_Molecular consequence    Zygosity  \n",
            "0                             Synonymous  Homozygous  \n",
            "\n",
            "Invitae Summary\n",
            "   Patient ID (MRN) Patient Name    Result  Gene   Transcript    Variant  \\\n",
            "0           43234.0    John Pork  POSITIVE  TSC2  NM_000548.3  c.1513C>T   \n",
            "1               NaN          NaN  NEGATIVE     -            -          -   \n",
            "\n",
            "  Protein Change Intron/Exon      Zygosity Classification  \n",
            "0      p.Arg505*     Exon 15  heterozygous     PATHOGENIC  \n",
            "1              -         NaN             -              -  \n",
            "\n",
            "Clinical Summary\n",
            "  Patient Name  Identification No.  \\\n",
            "0      Jon Doe               420.0   \n",
            "1          NaN                 NaN   \n",
            "\n",
            "                            Medical Prob description  \n",
            "0  Shwachman-Diamond Syndrome\\nloose stools- fat ...  \n",
            "1  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...  \n",
            "\n",
            "Research Summary\n",
            "  AUTO STATUS  IC No (MRN)        Name  \\\n",
            "0      Solved       1234.0  John Smith   \n",
            "1      Solved          NaN         NaN   \n",
            "\n",
            "                                           Phenotype Candidate gene (1)  \\\n",
            "0                   GDD with regression, Autism, FTT              CDK13   \n",
            "1  Dysmorphism, facial dysmorphism: arched eyebro...              KMT2D   \n",
            "\n",
            "  Transcript (1)     cDNA (1)  Protein (1)  \n",
            "0    NM_003718.5    c.2389G>C  p.Asp797His  \n",
            "1    NM_003482.3  c.12063dupG    p.T4022fs  \n"
          ]
        }
      ],
      "source": [
        "# Print dataframes\n",
        "print('Printing dataframes...')\n",
        "\n",
        "print('\\nLab Cases')\n",
        "print(lab_cases_df)\n",
        "\n",
        "print('\\nATM Summary')\n",
        "print(atm_structured_df)\n",
        "\n",
        "print('\\nInvitae Summary')\n",
        "print(invitae_summary_df)\n",
        "\n",
        "print('\\nClinical Summary')\n",
        "print(clinical_summary_df)\n",
        "\n",
        "print('\\nResearch Summary')\n",
        "print(research_summary_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 246,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DLRt0qwiIKBZ",
        "outputId": "cc56d155-61aa-44c5-ed47-87335776d09d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Standardising columns across dataframes...\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n",
            "Passed.\n"
          ]
        }
      ],
      "source": [
        "# Standardise columns across dataframes\n",
        "print('Standardising columns across dataframes...')\n",
        "\n",
        "# Lab Cases\n",
        "lab_cases_cols = {\n",
        "    'Number variants detected'    :'Var Count',\n",
        "    'Variant_1_gene'              :'Gene',\n",
        "    'Variant_1_HGVSg'             :'HGVSg',\n",
        "    'Variant_1_HGVSc'             :'HGVSc',\n",
        "    'Variant_1_HGVSp'             :'HGVSp',\n",
        "    'Variant_1_zygosity'          :'Zygosity',\n",
        "    'Variant_1_inheritance'       :'Inheritance',\n",
        "    'Variant_1_Validation_Status' :'Solved Status'\n",
        "}\n",
        "\n",
        "lab_cases_df.rename(columns = lab_cases_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# ATM Summary\n",
        "atm_structured_df_cols = {\n",
        "    'HGVS_Genomic_GRCh38/hg38'            :'HSVSg',\n",
        "    'HGVS_MANE Select_Transcript_RefSeq'  :'Transcript',\n",
        "    'HGVS_MANE Select_cDNA'               :'HGVSc',\n",
        "    'HGVS_MANE Select_protein'            :'HGVSp',\n",
        "    'HUGO gene symbol'                    :'Gene',\n",
        "}\n",
        "\n",
        "atm_structured_df.rename(columns = atm_structured_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# Invitae Summary\n",
        "invitae_summary_df_cols = {\n",
        "    'Patient ID (MRN)'  :'MRN',\n",
        "    'Patient Name'      :'Patient Name',\n",
        "    'Result'            :'Solved Status',\n",
        "    'Gene'              :'Gene',\n",
        "    'Transcript'        :'Transcript',\n",
        "    'Variant'           :'Variant',\n",
        "    'HGVSc'             :'HGVSc',\n",
        "    'Protein Change'    :'HGVSp',\n",
        "}\n",
        "\n",
        "invitae_summary_df.rename(columns = invitae_summary_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# Clinical Summary\n",
        "clinical_summary_df_cols = {\n",
        "    'Identification No.'        :'MRN',\n",
        "    'Medical Prob description'  :'Phenotype',\n",
        "    'Patient Name'              :'Patient Name'\n",
        "}\n",
        "\n",
        "clinical_summary_df.rename(columns = clinical_summary_df_cols, inplace = True)\n",
        "print('Passed.')\n",
        "\n",
        "# Research Summary\n",
        "research_summary_df_cols = {\n",
        "    'IC No (MRN)'         :'MRN',\n",
        "    'Name'                :'Patient Name',\n",
        "    'Candidate gene (1)'  :'Gene',\n",
        "    'Transcript (1)'      :'Transcript',\n",
        "    'cDNA (1)'            :'HGVSc',\n",
        "    'Protein (1)'         :'HGVSp',\n",
        "    'AUTO STATUS'         :'Solved Status'\n",
        "}\n",
        "\n",
        "research_summary_df.rename(columns = research_summary_df_cols, inplace = True)\n",
        "print('Passed.')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 247,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "6x-Sz1WP8kvV",
        "outputId": "dca32e9a-ca1e-40f9-db39-3da79294bf78"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Printing dataframes...\n",
            "\n",
            "Lab Cases\n",
            "                                           Phenotype    Gene  \\\n",
            "0  Developmental / Behavioural - Global developme...   PRRT2   \n",
            "1  Developmental / Behavioural - Global developme...   PRRT2   \n",
            "2  Developmental / Behavioural - Global developme...  STING1   \n",
            "3  Developmental / Behavioural - Global developme...  DYRK1A   \n",
            "4  Developmental / Behavioural - Delayed fine mot...   NTRK1   \n",
            "5  Developmental / Behavioural - Global developme...  DYRK1A   \n",
            "6  Developmental / Behavioural - Global developme...  DYRK1A   \n",
            "\n",
            "                         HGVSg                    HGVSc  \\\n",
            "0          chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "1          chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2  NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "4          chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "5          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "6          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "\n",
            "                           HGVSp      Solved Status  \n",
            "0  NP_660282.2:p.Arg217ProfsTer8  Confirmed present  \n",
            "1  NP_660282.2:p.Arg217ProfsTer8  Confirmed present  \n",
            "2        NP_938023.1:p.Val155Met  Confirmed present  \n",
            "3     NP_001334650.1:p.Trp146Ter  Confirmed present  \n",
            "4        NP_002520.2:p.Cys656Gly  Confirmed present  \n",
            "5     NP_001334650.1:p.Trp146Ter  Confirmed present  \n",
            "6     NP_001334650.1:p.Trp146Ter  Confirmed present  \n",
            "\n",
            "ATM Summary\n",
            "0                        HSVSg   Transcript      HGVSc         HGVSp  \\\n",
            "0  NC_000011.10:g.108281168G>A  NM_000051.3  c.3576G>A  p.(Lys1192=)   \n",
            "\n",
            "0 HGVS_MANE Select_Exon HGVS_MANE Select_Molecular consequence    Zygosity  \n",
            "0                 24/53                             Synonymous  Homozygous  \n",
            "\n",
            "Invitae Summary\n",
            "       MRN Patient Name Solved Status  Gene   Transcript    Variant  \\\n",
            "0  43234.0    John Pork      POSITIVE  TSC2  NM_000548.3  c.1513C>T   \n",
            "1      NaN          NaN      NEGATIVE     -            -          -   \n",
            "\n",
            "       HGVSp Intron/Exon      Zygosity Classification  \n",
            "0  p.Arg505*     Exon 15  heterozygous     PATHOGENIC  \n",
            "1          -         NaN             -              -  \n",
            "\n",
            "Clinical Summary\n",
            "  Patient Name    MRN                                          Phenotype\n",
            "0      Jon Doe  420.0  Shwachman-Diamond Syndrome\\nloose stools- fat ...\n",
            "1          NaN    NaN  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...\n",
            "\n",
            "Research Summary\n",
            "  Solved Status     MRN Patient Name  \\\n",
            "0        Solved  1234.0   John Smith   \n",
            "1        Solved     NaN          NaN   \n",
            "\n",
            "                                           Phenotype   Gene   Transcript  \\\n",
            "0                   GDD with regression, Autism, FTT  CDK13  NM_003718.5   \n",
            "1  Dysmorphism, facial dysmorphism: arched eyebro...  KMT2D  NM_003482.3   \n",
            "\n",
            "         HGVSc        HGVSp  \n",
            "0    c.2389G>C  p.Asp797His  \n",
            "1  c.12063dupG    p.T4022fs  \n"
          ]
        }
      ],
      "source": [
        "# Print dataframes after renaming\n",
        "print('Printing dataframes...')\n",
        "\n",
        "print('\\nLab Cases')\n",
        "print(lab_cases_df)\n",
        "\n",
        "print('\\nATM Summary')\n",
        "print(atm_structured_df)\n",
        "\n",
        "print('\\nInvitae Summary')\n",
        "print(invitae_summary_df)\n",
        "\n",
        "print('\\nClinical Summary')\n",
        "print(clinical_summary_df)\n",
        "\n",
        "print('\\nResearch Summary')\n",
        "print(research_summary_df)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def standardise_dataframe(df, columns):\n",
        "  \"\"\"\n",
        "  Converts unstandardised dataframe to standardised dataframe.\n",
        "\n",
        "  Args:\n",
        "    df (pd.DataFrame): Input dataframe.\n",
        "    columns (list): List of standardised column names.\n",
        "\n",
        "  Returns:\n",
        "    standardised_df (pd.DataFrame): Standardized dataframe.\n",
        "  \"\"\"\n",
        "  standardised_df = df.reindex(columns = columns, fill_value = np.nan)\n",
        "  return standardised_df\n",
        "\n",
        "print('Standardising all dataframes...')\n",
        "\n",
        "lab_cases_df_standard = standardise_dataframe(lab_cases_df, standard_columns)\n",
        "atm_structured_df_standard = standardise_dataframe(atm_structured_df, standard_columns)\n",
        "invitae_summary_df_standard = standardise_dataframe(invitae_summary_df, standard_columns)\n",
        "clinical_summary_df_standard = standardise_dataframe(clinical_summary_df, standard_columns)\n",
        "research_summary_df_standard = standardise_dataframe(research_summary_df, standard_columns)\n",
        "\n",
        "print('Standardisation complete.')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "qGfg0F1cxdkj",
        "outputId": "d2c009e1-9c89-4dfb-f122-1197106cd36f"
      },
      "execution_count": 248,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Standardising all dataframes...\n",
            "Standardisation complete.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Print dataframes to verify\n",
        "print('Individual dataframes:')\n",
        "print('\\nLab Cases:')\n",
        "print(lab_cases_df_standard.head())\n",
        "\n",
        "print('\\nATM:')\n",
        "print(atm_structured_df_standard.head())\n",
        "\n",
        "print('\\nInvitae:')\n",
        "print(invitae_summary_df_standard.head())\n",
        "\n",
        "print('\\nClinical:')\n",
        "print(clinical_summary_df_standard.head())\n",
        "\n",
        "print('\\nResearch:')\n",
        "print(research_summary_df_standard.head())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lrq6O3yrxlns",
        "outputId": "2112bf78-a534-4d0e-e84d-9aab2118e08f"
      },
      "execution_count": 249,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Individual dataframes:\n",
            "\n",
            "Lab Cases:\n",
            "   MRN  Patient Name                                          Phenotype  \\\n",
            "0  NaN           NaN  Developmental / Behavioural - Global developme...   \n",
            "1  NaN           NaN  Developmental / Behavioural - Global developme...   \n",
            "2  NaN           NaN  Developmental / Behavioural - Global developme...   \n",
            "3  NaN           NaN  Developmental / Behavioural - Global developme...   \n",
            "4  NaN           NaN  Developmental / Behavioural - Delayed fine mot...   \n",
            "\n",
            "       Solved Status    Gene  Transcript  Variant  \\\n",
            "0  Confirmed present   PRRT2         NaN      NaN   \n",
            "1  Confirmed present   PRRT2         NaN      NaN   \n",
            "2  Confirmed present  STING1         NaN      NaN   \n",
            "3  Confirmed present  DYRK1A         NaN      NaN   \n",
            "4  Confirmed present   NTRK1         NaN      NaN   \n",
            "\n",
            "                         HGVSg                    HGVSc  \\\n",
            "0          chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "1          chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2  NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3          chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "4          chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "\n",
            "                           HGVSp  \n",
            "0  NP_660282.2:p.Arg217ProfsTer8  \n",
            "1  NP_660282.2:p.Arg217ProfsTer8  \n",
            "2        NP_938023.1:p.Val155Met  \n",
            "3     NP_001334650.1:p.Trp146Ter  \n",
            "4        NP_002520.2:p.Cys656Gly  \n",
            "\n",
            "ATM:\n",
            "0  MRN  Patient Name  Phenotype  Solved Status  Gene   Transcript  Variant  \\\n",
            "0  NaN           NaN        NaN            NaN   NaN  NM_000051.3      NaN   \n",
            "\n",
            "0  HGVSg      HGVSc         HGVSp  \n",
            "0    NaN  c.3576G>A  p.(Lys1192=)  \n",
            "\n",
            "Invitae:\n",
            "       MRN Patient Name  Phenotype Solved Status  Gene   Transcript  \\\n",
            "0  43234.0    John Pork        NaN      POSITIVE  TSC2  NM_000548.3   \n",
            "1      NaN          NaN        NaN      NEGATIVE     -            -   \n",
            "\n",
            "     Variant  HGVSg  HGVSc      HGVSp  \n",
            "0  c.1513C>T    NaN    NaN  p.Arg505*  \n",
            "1          -    NaN    NaN          -  \n",
            "\n",
            "Clinical:\n",
            "     MRN Patient Name                                          Phenotype  \\\n",
            "0  420.0      Jon Doe  Shwachman-Diamond Syndrome\\nloose stools- fat ...   \n",
            "1    NaN          NaN  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...   \n",
            "\n",
            "   Solved Status  Gene  Transcript  Variant  HGVSg  HGVSc  HGVSp  \n",
            "0            NaN   NaN         NaN      NaN    NaN    NaN    NaN  \n",
            "1            NaN   NaN         NaN      NaN    NaN    NaN    NaN  \n",
            "\n",
            "Research:\n",
            "      MRN Patient Name                                          Phenotype  \\\n",
            "0  1234.0   John Smith                   GDD with regression, Autism, FTT   \n",
            "1     NaN          NaN  Dysmorphism, facial dysmorphism: arched eyebro...   \n",
            "\n",
            "  Solved Status   Gene   Transcript  Variant  HGVSg        HGVSc        HGVSp  \n",
            "0        Solved  CDK13  NM_003718.5      NaN    NaN    c.2389G>C  p.Asp797His  \n",
            "1        Solved  KMT2D  NM_003482.3      NaN    NaN  c.12063dupG    p.T4022fs  \n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 250,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xQd3sGenNIzt",
        "outputId": "9e7b6186-d754-45a4-92ad-0f4c6b8f27da"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Combining dataframes...\n",
            "All dataframes combined successfully.\n",
            "Combined dataframe:\n",
            "        MRN Patient Name                                          Phenotype  \\\n",
            "0       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "1       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "2       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "3       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "4       NaN          NaN  Developmental / Behavioural - Delayed fine mot...   \n",
            "5       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "6       NaN          NaN  Developmental / Behavioural - Global developme...   \n",
            "7       NaN          NaN                                                NaN   \n",
            "8   43234.0    John Pork                                                NaN   \n",
            "9       NaN          NaN                                                NaN   \n",
            "10    420.0      Jon Doe  Shwachman-Diamond Syndrome\\nloose stools- fat ...   \n",
            "11      NaN          NaN  6 weeks pregnant\\n\\nHistory of:\\nVSD\\nAnal atr...   \n",
            "12   1234.0   John Smith                   GDD with regression, Autism, FTT   \n",
            "13      NaN          NaN  Dysmorphism, facial dysmorphism: arched eyebro...   \n",
            "\n",
            "        Solved Status    Gene   Transcript    Variant  \\\n",
            "0   Confirmed present   PRRT2          NaN        NaN   \n",
            "1   Confirmed present   PRRT2          NaN        NaN   \n",
            "2   Confirmed present  STING1          NaN        NaN   \n",
            "3   Confirmed present  DYRK1A          NaN        NaN   \n",
            "4   Confirmed present   NTRK1          NaN        NaN   \n",
            "5   Confirmed present  DYRK1A          NaN        NaN   \n",
            "6   Confirmed present  DYRK1A          NaN        NaN   \n",
            "7                 NaN     NaN  NM_000051.3        NaN   \n",
            "8            POSITIVE    TSC2  NM_000548.3  c.1513C>T   \n",
            "9            NEGATIVE       -            -          -   \n",
            "10                NaN     NaN          NaN        NaN   \n",
            "11                NaN     NaN          NaN        NaN   \n",
            "12             Solved   CDK13  NM_003718.5        NaN   \n",
            "13             Solved   KMT2D  NM_003482.3        NaN   \n",
            "\n",
            "                          HGVSg                    HGVSc  \\\n",
            "0           chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "1           chr16:g.29813703dup     NM_145239.3:c.649dup   \n",
            "2   NC_000005.10:g.139480847C>T     NM_198282.4:c.463G>A   \n",
            "3           chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "4           chr1:g.156879282T>G    NM_002529.4:c.1966T>G   \n",
            "5           chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "6           chr21:g.37480775G>A  NM_001347721.2:c.438G>A   \n",
            "7                           NaN                c.3576G>A   \n",
            "8                           NaN                      NaN   \n",
            "9                           NaN                      NaN   \n",
            "10                          NaN                      NaN   \n",
            "11                          NaN                      NaN   \n",
            "12                          NaN                c.2389G>C   \n",
            "13                          NaN              c.12063dupG   \n",
            "\n",
            "                            HGVSp  \n",
            "0   NP_660282.2:p.Arg217ProfsTer8  \n",
            "1   NP_660282.2:p.Arg217ProfsTer8  \n",
            "2         NP_938023.1:p.Val155Met  \n",
            "3      NP_001334650.1:p.Trp146Ter  \n",
            "4         NP_002520.2:p.Cys656Gly  \n",
            "5      NP_001334650.1:p.Trp146Ter  \n",
            "6      NP_001334650.1:p.Trp146Ter  \n",
            "7                    p.(Lys1192=)  \n",
            "8                       p.Arg505*  \n",
            "9                               -  \n",
            "10                            NaN  \n",
            "11                            NaN  \n",
            "12                    p.Asp797His  \n",
            "13                      p.T4022fs  \n"
          ]
        }
      ],
      "source": [
        "# Combining dataframes into a single dataframe\n",
        "print('\\nCombining dataframes...')\n",
        "combined_df = pd.concat([\n",
        "    lab_cases_df_standard,\n",
        "    atm_structured_df_standard,\n",
        "    invitae_summary_df_standard,\n",
        "    clinical_summary_df_standard,\n",
        "    research_summary_df_standard\n",
        "], ignore_index = True)\n",
        "\n",
        "print('All dataframes combined successfully.')\n",
        "\n",
        "print('Combined dataframe:')\n",
        "print(combined_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "u8dZt3bQKo7i",
        "outputId": "1a33e034-c6c2-40f9-fec0-7b310df52c86"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "### Query Dataframe ###\n",
            "Available categories: MRN, Patient Name, Phenotype, Solved Status, Gene, Transcript, Variant, HGVSg, HGVSc, HGVSp\n",
            "\n"
          ]
        }
      ],
      "source": [
        "def query(df):\n",
        "  \"\"\"\n",
        "  Queries dataframe based on user input.\n",
        "\n",
        "  Args:\n",
        "    df (pd.DataFrame): Input dataframe.\n",
        "\n",
        "  Returns:\n",
        "    query_result (pd.DataFrame): Query result.\n",
        "  \"\"\"\n",
        "  print('\\n### Query Dataframe ###')\n",
        "  print(f\"Available categories: {', '.join(df.columns)}\\n\")\n",
        "\n",
        "  try:\n",
        "    # User input category\n",
        "    category = input('Enter category: ').strip()\n",
        "    category_lower = category.lower() # convert to lower case for non-case sens.\n",
        "\n",
        "    # Check if category exists\n",
        "    categories_lower = [col.lower() for col in df.columns]\n",
        "    if category_lower not in categories_lower:\n",
        "      raise ValueError(f\"Invalid category '{category}'. Please choose from available categories.\")\n",
        "\n",
        "    # Find actual column name (case sens match)\n",
        "    category_actual = df.columns[categories_lower.index(category_lower)]\n",
        "\n",
        "    # User input item (case insensitive and substring search)\n",
        "    item = input(f\"Enter item in category '{category_actual}': \").strip()\n",
        "\n",
        "    # Query dataframe with case insensitive comparison\n",
        "    query_result = df[df[category_actual].astype(str).str.contains(item, case = False, na = False)]\n",
        "\n",
        "    if query_result.empty:\n",
        "      print(f\"No results found for item containing '{item}' in category '{category_actual}'.\")\n",
        "    else:\n",
        "      print(f\"\\nQuery Results for item ({len(query_result)} records(s) found):\")\n",
        "      return query_result\n",
        "\n",
        "  except ValueError as e:\n",
        "    print(f\"Error: {e}\")\n",
        "    return None\n",
        "\n",
        "query_result = query(combined_df)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "dz5s_DIsKz-Y"
      },
      "outputs": [],
      "source": [
        "def export_to_excel(df, output_path):\n",
        "    \"\"\"\n",
        "    Exports the given DataFrame to an Excel file.\n",
        "\n",
        "    Args:\n",
        "        df (pd.DataFrame): The DataFrame to export.\n",
        "        output_path (str): Path where the Excel file will be saved.\n",
        "\n",
        "    Returns:\n",
        "        None\n",
        "    \"\"\"\n",
        "    try:\n",
        "        df_filled = df.fillna('NA')\n",
        "        df_filled.to_excel(output_path, index=False)\n",
        "        print(f\"Data successfully exported to {output_path}\")\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred while exporting to Excel: {e}\")\n",
        "\n",
        "output_file = f\"{local_path}Results/combined_data.xlsx\"\n",
        "export_to_excel(query_result, output_file)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "mount_file_id": "1EE8ToMNZVkhIRr6N6eiUAO9sxYAe6QwX",
      "authorship_tag": "ABX9TyPVrSD0dycuS3xgNnOuFDFT",
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}